template/starlark/convert_test.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2020 Target Brands, Inc. All rights reserved.
+//
+// Use of this source code is governed by the LICENSE file in this repository.
+
+package starlark
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/go-vela/types/raw"
+	"github.com/stretchr/testify/assert"
+	"go.starlark.net/starlark"
+)
+
+func TestNative_Render_convertTemplateVars(t *testing.T) {
+	// setup types
+	tags := starlark.Tuple(nil)
+	tags = append(tags, starlark.String("latest"))
+	tags = append(tags, starlark.String("1.14"))
+	tags = append(tags, starlark.String("1.15"))
+
+	commands := starlark.NewDict(16)
+	err := commands.SetKey(starlark.String("test"), starlark.String("go test ./..."))
+	assert.NoError(t, err)
+
+	strWant := starlark.NewDict(0)
+	err = strWant.SetKey(starlark.String("pull"), starlark.String("always"))
+	assert.NoError(t, err)
+
+	arrayWant := starlark.NewDict(0)
+	err = arrayWant.SetKey(starlark.String("tags"), tags)
+	assert.NoError(t, err)
+
+	mapWant := starlark.NewDict(0)
+	err = mapWant.SetKey(starlark.String("commands"), commands)
+	assert.NoError(t, err)
+
+	tests := []struct {
+		name string
+		args map[string]interface{}
+		want *starlark.Dict
+	}{
+		{
+			name: "test for a user passed string",
+			args: map[string]interface{}{"pull": "always"},
+			want: strWant,
+		},
+		{
+			name: "test for a user passed array",
+			args: map[string]interface{}{"tags": []string{"latest", "1.14", "1.15"}},
+			want: arrayWant,
+		},
+		{
+			name: "test for a user passed map",
+			// nolint // ignore line length
+			args: map[string]interface{}{"commands": map[string]string{"test": "go test ./..."}},
+			want: mapWant,
+		}}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := convertTemplateVars(tt.args)
+			assert.NoError(t, err)
```

Comment:
tl;dr 

* Are we ok with using a non-native Go package in our testing here?
* Are we ok with this opening up some inconsistencies in the way we write tests?

Context:

A long, **long** time ago we made a decision to **not** use any testing frameworks or helpers unless they were bundled in the native packages provided by Go.

My memory is a bit foggy, but I believe this decision was made back when this product had a different name and the source code was only available internally to Target.

I'm very familiar with [stretchr/testify](https://github.com/stretchr/testify) and why it's so widely used in the Go community so I'm not trying to debate it's value.

More so, I'm curious, are we planning to deviate from the above decision?

From a personal perspective, I often favor consistency above all else, because if I'm new to the Vela codebase, or just Go in general, it can be difficult to understand why certain things are done the way they are if they are accomplishing the same thing from a functional perspective.

As an example, here we are using the `assert` package from [stretchr/testify](https://github.com/stretchr/testify) but then later on in this file, we're back to using the native Go package with `t.Errorf`:

https://github.com/go-vela/compiler/blob/tmpl/starlark/template/starlark/convert_test.go#L121-L124

Reply:
> Are we ok with using a non-native Go package in our testing here?

I'm somewhat indifferent if we utilize testify and/or the default testing packages.

> Are we ok with this opening up some inconsistencies in the way we write tests?

I agree that we would ideally have consistency between testing files.. Would also point out the current inconsistency between table tests vs non table tests. Ideally migrating to some sort of baseline code generation like [this](https://code.visualstudio.com/docs/languages/go#_test) would help reduce the likelihood of differences between files/repos.

Reply:
I would support that notion to stay consistent. It looks like we're only using `assert.NoError()` as well (i could have missed something, though).

Reply:
Just pushed another commit to remove the testify package.

Reply:
I think the main point on why it was used, was in a lot of our testing we use `_` and throw errors away. That ended up being a pain point while we were developing and we needed a quick fix to see those errors. We probably don't need to have that external library in there and I agree on the consistency points. 


I would describe those `asserts` more as just there to help us debug and should have been cleaned up but were missed before we converted from draft to RFR.

Reply:
I fear my initial questions may have been too strongly worded 😅 

My primary intention for asking the questions was hoping to open the floor for discussions on the subject.

But, they did seem to accomplish part of what I intended especially based off @JordanSussman comments:

> I agree that we would ideally have consistency between testing files.. Would also point out the current inconsistency between table tests vs non table tests. Ideally migrating to some sort of baseline code generation like [this](https://code.visualstudio.com/docs/languages/go#_test) would help reduce the likelihood of differences between files/repos.

I agree with that and love the idea of picking a method of generating them like the link you provided.

Similar to the purpose of the linter we have setup, ideally that would help introduce more consistency and stability into the way we write code, specifically for tests in this case.

I think most, if not all of us, on our side are already using VScode as our IDE so that's a step in the right direction! 👍

Reply:
@kneal - I'd argue the value of `github.com/google/go-cmp/cmp` is a worthwhile import since the standard of `if !reflect.DeepEqual(got, want) ..` makes comparing the actual diff near impossible. Any objections to adding that back?

Reply:
I'm not opposed, it does have some issues with certain reflections. [kr/pretty](https://github.com/kr/pretty) might handle them better but the times I have seen the reflection issues are pretty rare. So, I think either would be a viable option if we wanted improved diffs over the reflect package.

Reply:
I haven't forgotten about this 👍 

But I was waiting to see where the dust settled on the github.com/google/go-cmp vs github.com/kr/pretty

@JordanSussman @kneal thoughts?

Reply:
My initial vote is for github.com/google/go-cmp since it appears to have more activity compared to github.com/kr/pretty.

Reply:
Works for me!

Reply:
I just added the github.com/google/go-cmp package back, we should consider a separate PR for updating other tests to utilize that as well.

@jbrockopp - ready for another review.